lib/src/sdk/sdk_package_config.dart

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// The top level structure of an `sdk_packages.yaml` file.
+class SdkPackageConfig {
+  /// The name of the SDK this configuration is for.
+  ///
+  /// SDKs should validate the config is for them on load.
+  final String sdk;
+
+  /// All the packages vendored by this SDK.
+  final List<SdkPackage> packages;
+
+  SdkPackageConfig(this.sdk, this.packages);
+
+  // Note: yaml has `Object?` keys.
+  SdkPackageConfig.fromMap(Map<Object?, Object?> map)
+      : sdk = map['sdk'] as String,
+        packages = [
+          for (var package in map['packages'] as List<Object?>)
+            SdkPackage.fromMap(package as Map<Object?, Object?>),
+        ];
+
+  Map<String, Object?> toMap() => {
+        'sdk': sdk,
+        'packages': [
+          for (var package in packages) package.toMap(),
+        ],
+      };
+}
+
+/// The structure for each `packages` entry in an `sdk_packages.yaml` file.
+class SdkPackage {
+  /// The name of the package.
+  final String name;
+
+  /// The path to the root of this package relative to the root of the SDK.
```

Comment:
> For configuration files, I'll usually make relative paths be relative to the configuration file itself. That doesn't require a second concept to interpret the file, the file is self-contained (but it does require dragging around the location of the file to where the contents are interpreted).

I would usually do that, but related to the next note, this file won't _actually_ be developed in a place where the relative paths would make sense. It only makes sense for the final built SDK, and so I think it is easier to speak in terms of that built SDK (that way this file can move for instance).

> I assume this is relative to the installed SDK root, so this file is generated as part of building the SDK release, where we copy the files from wherever they live in the SDK sources to a place in the deployment format, right?

Pretty much, but the file will likely be checked in as source and just copied to the right location in the final SDK.

Reply:
In that case, this file can actually be the source of truth for *where* the package should be copied.
Which also supports using the SDK root as base, because that's what we're copying into.